### ISSUE TITLE
Discussion: store settings as INI files

### ISSUE THREAD

**Issue by virtualtam** [shaarli/Shaarli#489]
### On the current settings
The settings are currently loaded & written as PHP code, which I'd flag as (very) bad usage regarding the following criteria:
- configuration should use a human-readable format, not be tied to a given programming language
- entries should be proof-checked and sanitized before being used
- configuration should _definitely_ not be imported to _set global variables_ right away

Also, it's a bit weird to have to write configuration handling functions when most programming languages provide full-featured parsers out-of-the-box :p

### On INI files
From #471:

> INI configuration files are not suited for configuration with array

INI is a simple configuration file format that allows defining:
- sections, to group settings by usage/tool/category
- key/value pairs

How the _value_ part is used really depends on the needs, and is quite flexible:
- single numerical or alphanumerical value
- comma-separated lists that can receive further processing from the code

Additionally, several INIs may be combined, e.g. to set application default values, then override them with user preferences.

### Real-life example
INI / CFG files are widely used, I've seen them mostly holding user preference for games or desktop applications. INI is also the reference format for storing the configuration of most Python packaging, test and code quality tools.

If you take a look at the [UT2004](http://www.unrealadmin.org/server_ini_reference/ut2004) server configuration reference, it features:
- numeric values (integers, floats)
- booleans
- strings
- lists of values
- lists of key/value paors
- arrays

Some examples issued from it:
```ini
[IpDrv.MasterServerLink] 	
LANPort=11777
LANServerPort=10777
MasterServerList=(Address="ut2004master1.epicgames.com",Port=28902)
MasterServerList=(Address="ut2004master2.epicgames.com",Port=28902)
; [...]

[UnrealGame.DeathMatch]
; [...]
NamePrefixes[0]=Mr_
NamePrefixes[1]=
NamePrefixes[2]=The_Real_
NamePrefixes[3]=Evil_
NamePrefixes[4]=
```

### Shaarli example / suggestion

Here's a quick shot on how Shaarli configuration could be stored as multiple INI files:

Main `data/config.ini`:
- user- and group-readable
- user- and group-writable

```ini
[authentication]
hash = c5e415644bcd5755
login = toto
salt = abc123def456

[main]
title = My Suuper Shaarewares
titlelink = ?

[privacy]
open = false
hide_public_links = false

[raintpl]
tmp_dir = tmp/
tpl_dir = tpl/

[resources]
cache_dir = cache
data_dir = data
datastore = data/datastore.php

[security]
ban_after = 5
ban_duration = 1800
session_protection = true
redirector =

[updates]
check_interval = 10
enable = true
git_branch = stable

[plugins]
enabled = (addlink_toolbar,markdown,qrcode,archiveorg)

[plugin:my_active_plugin]
background = images/pink_fluffy_unicorn.png
header_color = #224433
```

Default `plugins/my_supah_plugin/defaults.ini`:
- user- and group-readable
- can hold plugin metadata
- provides good plugin defaults (if any)

```ini
[plugin:my_supah_plugin]
description = IMAH CODIN MAH PLUGIN
some_key = (some,bunch,of,values)
```

**Comment by Marsup** [comment]
Out of curiosity, why ini over say yaml ? Seems to me that yaml is at least as readable and handles the types you mentioned (with a better syntax I might add). It also supports nesting which might be useful sometimes...

**Comment by virtualtam** [comment]
> why ini over say yaml?

Several practical reasons:
- INI is supported by vanilla PHP, YAML is not
- there was a YAML proposal in #108, TL;DR: too many code added, increased complexity

Proper YAML support would require either including or referencing a third-party library with Composer, which would increase the complexity when installing. This could be done, as Composer is already used to handle test and code quality dependencies, but would break the KISS install process and require users to keep dependencies up-to-date.

> yaml is at least as readable

As a rule of thumb:
- INI => human-readable - application configuration, user settings
- YAML => human-readable - user data, declarative data structures
- XML => machine-generated, machine-readable - application configuration (Java), web-transferred data (Java), semantically constrained data
- JSON => similar to XML, just `s/Java/<hype language>/g` and append "wild free-form data structures"

> a better syntax
> supports nesting which might be useful

Yup, if you need a human-readable format to store simple data structures such as lists or dictionaries.

> sometimes

I think the key point is that as far as Shaarli settings' storage is concerned, we actually do not need the extra features and may stick with simple key/value pairs, hence this issue ^^

**Comment by Aldarone** [comment]
> This could be done, as Composer is already used to handle test and code quality dependencies, but would break the KISS install process and require users to keep dependencies up-to-date.

About that : The github release feature allows to upload and associate arbitrary files to a release. So the maintainers could tag a release, install the dependencies using composer and gzip the result for github. I think that's what the Wallabag people do.

**Comment by virtualtam** [comment]
Thanks @Aldarone for the suggestion :)

This only seem to apply to manually created releases though (tag & push a commit, create a release, add stuff to it), so would only benefit people getting Shaarli through GH releases, and users of the Docker images.

Documentation:
- [Creating releases](https://help.github.com/articles/creating-releases/)
- [Release API](https://developer.github.com/v3/repos/releases/#create-a-release)

I'd be in favour of using a dependency manager for 3rd party deps as it would help us strip the codebase down while benefiting from (more) robust libraries. Most PHP projects I can think of do this (or have come to), while it's a core feature of other languages (cf. Ruby gems & Python wheels).

@nodiscc & @ArthurHoaro , what do you think of it?

**Comment by nodiscc** [comment]
I have mixed thoughts on this, mainly because I would like to keep plain `git` commands as a working upgrade/downgrade/revert/test mechanism. Using composer or another dependency manager would mean, during upgrades, either

a) manually backing up `data/`, removing the main shaarli dir, downloading the new version zip, extracting it, restoring old `data/` (basically the "old" Shaarli method)..

b) running the build tools on my server to get a working installation - I find software that does this annoying, my server is not a development machine and I don't want to install every toolchain out there.

I'm not opposed to a "proper" build system using one or another dependency manager, but it needs to have the same simplicity as `git pull` for upgrades, at the very least.

**Comment by virtualtam** [comment]
@nodiscc I'm using git+composer to keep my [kanboard](https://github.com/fguillot/kanboard) installation up-to-date on a weekly basis, which mostly requires to:
- have composer somewhere in your `$PATH`, or `composer.phar` in a local folder
- `git pull` to get the PHP code up-to-date
- `composer update` to update PHP dependencies when needed

Best practices regarding dependency management are:
- keeping it dead simple for users
- ensuring the code is run in an environment it is compatible with (pinned dependencies)
- avoid versioning extra libraries and having to manage them

**Comment by Aldarone** [comment]
> One last thing how would it affect users running Shaarli on shared hosting (no shell access)?

Those can't use git anyway, so they're stuck at downloading a zipfile and sending the files to their host.

From there, two use cases : 
- Normal users who go from release to release. For them there will be the custom zip the Shaarli-team will provide with each Github Release.
- Users who want to try master. They'll need to get the master.zip, unpack it, `composer install` and send the result. Pretty much like the git users except they'll need composer on their computer.

**Comment by nodiscc** [comment]
Then ok for this  release method. Make sure we have some form of integrity/authenticity check for the downloads (git provides this through object hashes and signed commits).

**Comment by alexisju** [comment]
In my case, i use the shell and git on a shared hosting, but composer is not available... :(

**Comment by ArthurHoaro** [comment]
The discussion has a bit shifted here. I also have mixed feelings about composer, because it adds complexity and users don't necessarily know PHP's set of tools. Alda's suggestion is a good compromise, but anyway, Shaarli has one single PHP dependency: *RainTPL* which is more or less dead. If we need to add new dependencies, maybe we can discuss it in another issue?

About INI files, I have to admit that your example looks good, and I didn't know this syntax:
```ini
MasterServerList=(Address="ut2004master2.epicgames.com",Port=28902)
```
However, PHP support for INI files seems really poor:

  * apparently there is no function to write the INI file.
  * file reading with [`parse_ini_file()`](http://php.net/manual/en/function.parse-ini-file.php) isn't really good. The `MasterServerList` syntax fails, and everything else is handled as a string (before PHP 5.6.1 apparently) except booleans which become `1` or nothing.

In other words, we would have to have to write/find a lib to parse and write INI files. It kinda fails one of initial concern

> Also, it's a bit weird to have to write configuration handling functions when most programming languages provide full-featured parsers out-of-the-box

I think that JSON is the only format handled by "vanilla" PHP (it requires `php-json` which is bundled with most default PHP packages) with [`json_encode()`](http://php.net/manual/en/function.json-encode.php) and [`json_decode()`](http://php.net/manual/en/function.json-decode.php). Also, JSON is pretty much human readable.

So either we go with JSON, or INI/YAML with third party lib.

**Comment by mro** [comment]
yaml please. A sidenote: json is valid yaml. So consider json the yaml for machines.

**Comment by virtualtam** [comment]
@mro sadly nope, http://php.net/manual/en/yaml.installation.php

> This » PECL extension is not bundled with PHP.

This is one of the (oh so) frustrating things about PHP: all the good chunks are lying there, but they miss the final baking touch to nicely tie them together...

@ArthurHoaro JSON is human-readable if it's formatted so: machine-written files are usually one-liners. Most text editors and web browsers feature JSON unfolding, though.

> `php-json` which is bundled with most default PHP packages

I wouldn't bet on it, as it's a separate package in [Debian](https://packages.debian.org/jessie/php5-json) and [Ubuntu](http://packages.ubuntu.com/vivid/php5-json).

**Comment by ArthurHoaro** [comment]
On Debian:
> root@trantor:~# apt-get install php5
Reading package lists... Done
Building dependency tree
Reading state information... Done
The following extra packages will be installed:
> >  apache2 apache2-bin apache2-data apache2-utils libapache2-mod-php5 libapr1 libaprutil1 libaprutil1-dbd-sqlite3 libaprutil1-ldap liblua5.1-0 libonig2 libqdbm14 php5-cli php5-common **php5-json** php5-readline ssl-cert

But yeah, it really depends on the distribution choices...

**Comment by nodiscc** [comment]
I would favor the solution that adds the least possible complexity (both in release/upgrade methods and code complexity/LOC). JSON seems to fit the bill. Regarding human readability, `json.encode()` has a ` JSON_PRETTY_PRINT` option.

**Comment by nicolasdanelon** [comment]
I think we are good now. Why? Because WordPress and actual PrestaShop versions are storing important data in php files as constants. Everybody can read my shaarli log because of my bad nginx configuration. Something colud happens with ini files.

**Comment by virtualtam** [comment]
Thanks everyone for your feedback!
I'm closing this issue, as JSON would be the way to go if we were to refactor Shaarli's user settings management :)
